DOCSP-41621: upsert (#3089)
jenssegers/laravel-mongodb: docs/fundamentals/write-operations.txt

operation. This operation streamlines the task of updating a document or
inserting one if it does not exist.

Starting in v4.7, you can perform an upsert operation by using either of
the following methods:

> - Use the ``upsert()`` method. When you use this method,
>   you can perform a **batch upsert** to change or insert multiple documents
>   in one operation.
> 
> - Use the ``update()`` method and specify the ``upsert``

Reviewer: S: since you introduce this list with "using either of the following methods," I think it makes sense to start each bullet with the method itself. Or, update that intro text to something like "in the following ways:"
```suggestion
- ``upsert()``: When you use this method,
  you can perform a **batch upsert** to change or insert multiple documents
  in one operation.

- ``update()``: When you use this method, specify the ``upsert``
```

Reviewer: Applies to the query builder page as well

  or insert one document if no documents are matched. Only this upsert method
  is supported in versions v4.6 and earlier.

Upsert Method
~~~~~~~~~~~~~

> The ``upsert()`` query builder method accepts the following parameters:
> 
> - Array of fields and values that specify documents to update or insert.
> - Array of fields that uniquely identify documents in your first array parameter.
> - Array of fields to update if a matching document exists. If you omit
>   this parameter, {+odm-short+} updates all fields.

Reviewer 2: What do you think about having the names of the arguments? It seems easier to read. This is how all PHP functions and MongoDB operators are described in the official documentation

```suggestion
The ``upsert(array $values, array|string $uniqueBy, array|null $update)`` query builder method accepts the following parameters:

- ``$values``: Array of fields and values that specify documents to update or insert.
- ``$uniqueBy``: List of fields that uniquely identify documents in your first array parameter.
- ``$update``: Optional list of fields to update if a matching document exists. If you omit
  this parameter, {+odm-short+} updates all fields.
```

"list" is more specific than array to indicate that the keys are not relevant.

To specify an upsert in the ``upsert()`` method, set parameters
as shown in the following code example:

.. code-block:: php
   :copyable: false

>    YourModel::upsert(
>       [/* documents to update or insert */],
>       '/* unique field */', [/* fields to update */]
>    );

Reviewer 2: It looks like there is only 2 arguments. Splitting the lines.
```suggestion
   YourModel::upsert(
      [/* documents to update or insert */],
      '/* unique field */',
      [/* fields to update */],
   );
```

   );

Example
^^^^^^^

This example shows how to use the  ``upsert()``
method to perform an update or insert in a single operation. Click the
:guilabel:`{+code-output-label+}` button to see the resulting data changes if

> there is an ``'Angel Olson'`` document in the collection already:

Reviewer: S: clarify what "Angel Olson document" means

```suggestion
there is a document that has a ``performer`` value of ``'Angel Olson'`` in the collection already:
```

.. io-code-block::

   .. input:: /includes/fundamentals/write-operations/WriteOperationsTest.php
      :language: php
      :dedent:
      :start-after: begin model upsert
      :end-before: end model upsert

   .. output::
      :language: json
      :visible: false

      {
        "_id": "...",
        "performer": "Angel Olsen",
        "venue": "State Theatre",
        "genres": [
          "indie",
          "rock"
        ],
        "ticketsSold": 275,
        "updated_at": ...
      },
      {
        "_id": "...",
        "performer": "Darondo",
        "venue": "Cafe du Nord",
        "ticketsSold": 300,
        "updated_at": ...
      }

> In the ``'Angel Olsen'`` document, the ``venue`` field value is not

Reviewer: S: same comment as above
```suggestion
In the document that has a ``performer`` value of ``'Angel Olson'``, the ``venue`` field value is not
```
